p2p/server/server.go

```diff
@@ -356,12 +364,21 @@ func (s *Server) StreamRequest(
 	defer cancel()
 	stream, err := s.streamRequest(ctx, pid, req, extraProtocols...)
```

Comment:
Revisiting this part: with this updated code, `ServerError` is going to be missed. `streamRequest` only writes the request into the stream but doesn't read the response. The idea here is that the `callback` may choose to read the data in streaming mode interpreting it as it goes. The actual response is decoded in the `callback` using `ReadResponse` function, which then can return a `ServerError`, which needs to be included in `clientServerError` metric. An I/O error during the `callback` can happen too, and then it needs to be included in `clientFailed` metric. Also, `clientLatency.Observe` / `clientLatenyFailure.Observe` also needs to be called in case of an error returned from the `callback`.

Aside from that, although you mentioned it previously, I may have forgotten some details: why prefer `errors.Is` over `errors.As`? Both methods follow the error chain and the difference is that we use `Is` when we don't need a target error instance to be extracted from the chain. The `ServerError.Is` method is needed so that any `&ServerError{...}` is considered a `ServerError`, which is important as it can carry a server-generated message, so having a `ServerError` variable is not convenient (we could wrap `ServerError` value instead but it's a bit inconvenient). The `ServerError.Is` is used both by `errors.Is` and `errors.As` in a similar manner. Pls check this example: https://goplay.tools/snippet/pOIGh5POHbB

Reply:
> why prefer `errors.Is` over `errors.As`

I don't. I said that by using `error.As` we can get rid of the (imo wrongly implemented) `Is` method on `ServerError`.

> The ServerError.Is method is needed so that any &ServerError{...} is considered a ServerError

Right and I have a problem with that 🙂 take this as example:

```golang
err1 := &ServerError{msg: "first error"}
err2 := &ServerError{msg: "second error"}

errors.Is(err1, err2) // returns true - which imo it should not
```

> which is important as it can carry a server-generated message, so having a ServerError variable is not convenient

it makes no difference:

```go
errors.Is(err, &ServerError{})
```

vs
 
```go
sErr := &ServerError{}
errors.Is(err, sErr)
```

In both cases a new variable needs to be created... It boils down to semantics - `error.Is` means equality to me, not equivalence. This is what godoc says about it:

```
Is reports whether any error in err's tree matches target.

The tree consists of err itself, followed by the errors obtained by repeatedly calling its Unwrap() error or Unwrap() []error method. When err wraps multiple errors, Is examines err followed by a depth-first traversal of its children.
```

in the case of `ServerError.Is` our implementation behaves in my opinion incorrectly - `&ServerError{}` is NOT wrapped in `err1 := &ServerError{msg: "first error"}` - they are not the same errors 🙂

Reply:
Yes, thanks, I see, the intent of the `ServerError.Is` method was to make it more convenient to deal with `ServerError`, namely so that `Is` could be used instead of `As`, but that's indeed not very idiomatic Go b/c `Is` is supposed to do by-value comparison of the errors, not treating the errors of the same type as the same error. I think the best thing would be just to use 
```
var ServerError = errors.New("server error")
```
and then wrap it together with the actual error message from the server; so that if you want to check for a `ServerError`, `Is` can be easily used, but different errors are not considered equal (per `Is`).  But I'd prefer to fix it in a separate PR as it is somewhat out of scope for this one.